consensus/hotstuff/verification/combined_verifier_v2.go

```diff
@@ -0,0 +1,142 @@
+}
+
+// NewCombinedVerifier creates a new combined verifier with the given dependencies.
+// - the hotstuff committee's state is used to retrieve the public keys for the staking signature;
+// - the staking tag is used to create hasher to verify staking signatures;
+// - the beacon tag is used to create hasher to verify random beacon signatures;
+// - the merger is used to combined & split staking & random beacon signatures;
+func NewCombinedVerifierV2(committee hotstuff.Committee, stakingTag string, beaconTag string, merger module.Merger) *CombinedVerifierV2 {
+	return &CombinedVerifierV2{
+		committee:      committee,
+		staking:        crypto.NewBLSKMAC(stakingTag),
+		beacon:         crypto.NewBLSKMAC(beaconTag),
+		keysAggregator: newStakingKeysAggregator(),
+		merger:         merger,
+	}
+}
+
+// VerifyVote verifies the validity of a combined signature from a vote.
+// Usually this method is only used to verify the proposer's vote, which is
+// the vote included in a block proposal.
+// TODO: return error only, because when the sig is invalid, the returned bool
+// can't indicate whether it's staking sig was invalid, or beacon sig was invalid.
+func (c *CombinedVerifierV2) VerifyVote(signer *flow.Identity, sigData []byte, block *model.Block) (bool, error) {
+
+	// create the to-be-signed message
+	msg := MakeVoteMessage(block.View, block.BlockID)
+
+	// split the two signatures from the vote
+	// TODO: move DecodeDoubleSig to merger.Split
+	stakingSig, beaconShare, err := signature.DecodeDoubleSig(sigData)
+	if err != nil {
+		return false, fmt.Errorf("could not split signature: %w", modulesig.ErrInvalidFormat)
+	}
+
+	dkg, err := c.committee.DKG(block.BlockID)
+	if err != nil {
+		return false, fmt.Errorf("could not get dkg: %w", err)
+	}
+
+	// verify each signature against the message
+	// TODO: check if using batch verification is faster (should be yes)
+	stakingValid, err := signer.StakingPubKey.Verify(stakingSig, msg, c.staking)
+	if err != nil {
+		return false, fmt.Errorf("internal error while verifying staking signature: %w", err)
+	}
+	if !stakingValid {
+		return false, fmt.Errorf("invalid staking sig")
+	}
+
+	// there is no beacon share, no need to verify it
+	if beaconShare == nil {
+		return true, nil
+	}
+
+	// if there is beacon share, there must be beacon public key
+	beaconPubKey, err := dkg.KeyShare(signer.NodeID)
+	if err != nil {
+		return false, fmt.Errorf("could not get random beacon key share for %x: %w", signer.NodeID, err)
+	}
+
+	beaconValid, err := beaconPubKey.Verify(beaconShare, msg, c.beacon)
+	if err != nil {
+		return false, fmt.Errorf("internal error while verifying beacon signature: %w", err)
+	}
+
+	if !beaconValid {
+		return false, fmt.Errorf("invalid beacon sig")
+	}
+	return true, nil
+}
+
+// VerifyQC verifies the validity of a combined signature on a quorum certificate.
+func (c *CombinedVerifierV2) VerifyQC(signers flow.IdentityList, sigData []byte, block *model.Block) (bool, error) {
```

Comment:
I think this method is not compatible with a QC from version V2: it still uses the old convention. Can we please update it, or stub it out? Having an unclear mixture of updated and legacy code makes it super hard to review.